### PR TITLE
Create foreman_user module

### LINKED ISSUE #200: Create foreman_user module
##### SUMMARY

I went trough all existing modules but haven't found a way how to create or manage a foreman user.
I'd expect that it would be best to create separate module for this.
Similiar functionality provides [os_user](https://docs.ansible.com/ansible/2.6/modules/os_user_module.html#os-user-module) module for OpenStack.
##### ISSUE TYPE
 - Feature Idea

### PR DESCRIPTION
**Pull request by ephracis**
Create a new module for managing users. In order to change password we
need to force the `password` field to be sent during POST but that
field is not included when doing GET, so we put it in `check_missing`.

I noticed, though, that the value for `check_missing` was a variable
from a separate loop which seems wrong. Instead I change it to get the
value from the incoming desired dict. This makes the module behave as
we want.

To ensure that the new user works we do an API request as the
newly created user. We do this again after changing password to ensure it worked.

Closes #200

### CODE REVIEW

File: modules/foreman_user.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    'auth_source': 'auth_source_id',
+    'timezone': 'timezone',
+    'locale': 'locale',
+    'roles': 'role_ids',
+    'locations': 'location_ids',
+    'organizations': 'organization_ids',
+}
+
+# List of allowed timezones
+timezone_list = [
```
Comment (sean797) [inline modules/foreman_user.py]: Can we pull these from apypie? It's should be exposed via https://github.com/Apipie/apypie/blob/master/apypie/param.py#L19

Comment (ephracis) [reply modules/foreman_user.py]: That would be awesome. I just need to figure out how to do that.

Comment (ephracis) [reply modules/foreman_user.py]: First of all, I don't understand Apypie well enough to know how to fetch that data from the server.

Secondly, as far as I can tell when checking `/apidoc/v2.json` is that the `validator` field just contains a human description and I'm not sure if that is stable enough for us to parse it and use it for validation of the module params.

Perhaps we could replace the hardcoded list in a later PR?

Comment (sean797) [reply modules/foreman_user.py]: I was just about to comment something similar. We're going to need changes to apypie. :+1:

File: module_utils/foreman_helper.py [hunk 2]
```diff
@@ -200,7 +200,7 @@ def update_resource(self, resource, old_entity, entity_dict, check_missing, chec
         if check_missing is not None:
             for key in check_missing:
                 if key in entity_dict and key not in volatile_entity.keys():
-                    volatile_entity[key] = new_value
+                    volatile_entity[key] = entity_dict[key]
```
Comment (evgeni) [inline module_utils/foreman_helper.py]: oups. :)
Good catch.

**Comment by mdellweg** [comment]
This probably also needs a rebase and the XDG_CACHE_HOME thing

**Comment by ephracis** [comment]
> This probably also needs a rebase and the XDG_CACHE_HOME thing

Done!

File: plugins/modules/foreman_user.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+  foreman_user:
+    name: test
+    user_password: newp@ss
+
+'''
+
+RETURN = ''' # '''
+
+from ansible.module_utils.foreman_helper import (
+    sanitize_entity_dict,
```
Comment (mdellweg) [inline plugins/modules/foreman_user.py]: you know ...

File: modules/foreman_user.py (hunk 1 above)
Comment (evgeni) [reply modules/foreman_user.py]: Can you please open an issue against apypie? The validators can use some love indeed ;)

Comment (ephracis) [reply modules/foreman_user.py]: https://github.com/Apipie/apypie/issues/27

File: plugins/modules/foreman_user.py (hunk 3 above)
Comment (ephracis) [reply plugins/modules/foreman_user.py]: Yeah I know... ;)

**Comment by evgeni** [comment]
Argh, seems my client didn't post the actual question.

Does the `if mail not in entity_dict` part work also when `entity` is `None` as when the user is freshly created? Or is `mail` mandatory for new creations?

**Comment by evgeni** [comment]
(it is from the description, wonder if we can make that more explicit in the code)

**Comment by ephracis** [comment]
> (it is from the description, wonder if we can make that more explicit in the code)

Yeah I was thinking that myself.  But it's kinda hard to do since we need to first check if the user exists before we know whether some params are required or not.

So I think the best and most boring solution would be to just document the requirement and let the error from Foreman bleed through the module. Maybe in the future we could improve it further.

File: plugins/modules/foreman_user.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    'auth_source': 'auth_source_id',
+    'timezone': 'timezone',
+    'locale': 'locale',
+    'roles': 'role_ids',
+    'locations': 'location_ids',
+    'organizations': 'organization_ids',
+}
+
+# List of allowed timezones
+timezone_list = [
```
Comment (mdellweg) [inline plugins/modules/foreman_user.py]: Are these lists (until they are generated) better suited to live somewhere in module_utils?

Comment (ephracis) [reply plugins/modules/foreman_user.py]: If we would use them in multiple modules they would be. But I'm not sure we will do that anytime soon. Do we want to stuff a lot of into the utils that is only used in a single module?

I vote for keeping stuff in modules, and then move into module_utils when it needs to be shared between two or more modules. What do you guys think?

File: test/test_playbooks/user.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
+---
+- hosts: fixtures
+  gather_facts: false
+  tasks:
+  - name: Load server config
+    include_vars:
```
Comment (mdellweg) [inline test/test_playbooks/user.yml]: I guess, you need to adjust that now.
(Directory changed...)

**Comment by mdellweg** [comment]
@ephracis Thank you

**Pull request merged**